derive/src/compile_bytecode.rs

```diff
@@ -0,0 +1,135 @@
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use quote::quote;
+use rustpython_compiler::{bytecode::CodeObject, compile};
+use std::env;
+use std::fs;
+use std::path::{Path, PathBuf};
+use syn::parse::{Parse, ParseStream, Result as ParseResult};
+use syn::{self, parse2, Ident, Lit, LitByteStr, Meta, MetaList, NestedMeta, Token};
+
+struct BytecodeConst {
+    ident: Ident,
+    meta: MetaList,
+}
+
+impl BytecodeConst {
+    fn compile(&self, manifest_dir: &Path) -> Result<CodeObject, Diagnostic> {
+        let meta = &self.meta;
+
+        let mut source_path = None;
+        let mut mode = None;
+        let mut source_lit = None;
+
+        for meta in &meta.nested {
+            match meta {
+                NestedMeta::Literal(lit) => source_lit = Some(lit),
+                NestedMeta::Meta(Meta::NameValue(name_value)) => {
+                    if name_value.ident == "mode" {
+                        mode = Some(match &name_value.lit {
+                            Lit::Str(s) => match s.value().as_str() {
+                                "exec" => compile::Mode::Exec,
+                                "eval" => compile::Mode::Eval,
+                                "single" => compile::Mode::Single,
+                                _ => bail_span!(s, "mode must be exec, eval, or single"),
+                            },
+                            _ => bail_span!(name_value.lit, "mode must be a string"),
+                        })
+                    } else if name_value.ident == "source_path" {
+                        source_path = Some(match &name_value.lit {
+                            Lit::Str(s) => s.value(),
+                            _ => bail_span!(name_value.lit, "source_path must be string"),
+                        })
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        let source = if meta.ident == "file" {
+            let path = match source_lit {
+                Some(Lit::Str(s)) => s.value(),
+                _ => bail_span!(source_lit, "Expected string literal for path to file()"),
+            };
+            let path = manifest_dir.join(path);
+            fs::read_to_string(&path)
+                .map_err(|err| err_span!(source_lit, "Error reading file {:?}: {}", path, err))?
+        } else if meta.ident == "source" {
+            match source_lit {
+                Some(Lit::Str(s)) => s.value(),
+                _ => bail_span!(source_lit, "Expected string literal for source()"),
+            }
+        } else {
+            bail_span!(meta.ident, "Expected either 'file' or 'source'")
+        };
+
+        compile::compile(
+            &source,
+            &mode.unwrap_or(compile::Mode::Exec),
+            source_path.unwrap_or_else(|| "frozen".to_string()),
+        )
+        .map_err(|err| err_span!(source_lit, "Compile error: {}", err))
+    }
+}
+
+impl Parse for BytecodeConst {
+    /// Parse the form `static ref IDENT = metalist(...);`
```

Comment:
I suggest to use a normal macro to introduce a marker token for the second stage.

First, use a macro something like this: `include_file_as_bytecode!("my_file.py")` which will expand to `compile_file_sentinel "my_file.py"` then the matching parse can use the `compile_file_sentinel` as a token to match on, so that users of the macro do not need to use `static ref IDENT ...`, but this will then also work: `let bytecode = include_file_as_bytecode!("other_file");`!

Would this be possible? Did I explain this idea well enough?

Reply:
I would have done this in the first place, but the problem is that procedural macros can't be used in expression position. There is a way to fix that, see: https://github.com/dtolnay/proc-macro-hack, but I wasn't sure if that was necessary. I suppose we could have `rustpython_vm`  be the declaration crate? Should I try that?

Reply:
Yes, that would be good if the resulting macro is easier to understand.